tls/s2n_tls13_key_schedule.c

```diff
@@ -0,0 +1,314 @@
+            &secret, purpose_iv, &s2n_zero_length_context, &iv));
+
+    bool is_sending_secret = (mode == conn->mode);
+    if (is_sending_secret) {
+        RESULT_GUARD_POSIX(cipher->set_encryption_key(session_key, &key));
+    } else {
+        RESULT_GUARD_POSIX(cipher->set_decryption_key(session_key, &key));
+    }
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#section-5.3
+     *# Each sequence number is
+     *# set to zero at the beginning of a connection and whenever the key is
+     *# changed; the first record transmitted under a particular traffic key
+     *# MUST use sequence number 0.
+     */
+    RESULT_GUARD(s2n_zero_sequence_number(conn, mode));
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_client_key_schedule(struct s2n_connection *conn)
+{
+    message_type_t message_type = s2n_conn_get_current_message_type(conn);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.1
+     *#                              START <----+
+     *#               Send ClientHello |        | Recv HelloRetryRequest
+     *#          [K_send = early data] |        |
+     */
+    if (message_type == CLIENT_HELLO
+            && conn->early_data_state == S2N_EARLY_DATA_REQUESTED) {
+        K_send(conn, S2N_EARLY_SECRET);
+    }
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.1
+     *#                                v        |
+     *#           /                 WAIT_SH ----+
+     *#           |                    | Recv ServerHello
+     *#           |                    | K_recv = handshake
+     */
+    if (message_type == SERVER_HELLO) {
+        K_recv(conn, S2N_HANDSHAKE_SECRET);
+    }
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.1
+     *#       Can |                    V
+     *#      send |                 WAIT_EE
+     *#     early |                    | Recv EncryptedExtensions
+     *#      data |           +--------+--------+
+     *#           |     Using |                 | Using certificate
+     *#           |       PSK |                 v
+     *#           |           |            WAIT_CERT_CR
+     *#           |           |        Recv |       | Recv CertificateRequest
+     *#           |           | Certificate |       v
+     *#           |           |             |    WAIT_CERT
+     *#           |           |             |       | Recv Certificate
+     *#           |           |             v       v
+     *#           |           |              WAIT_CV
+     *#           |           |                 | Recv CertificateVerify
+     *#           |           +> WAIT_FINISHED <+
+     *#           |                  | Recv Finished
+     *#           \                  | [Send EndOfEarlyData]
+     *#                              | K_send = handshake
+     */
+    if ((message_type == SERVER_FINISHED && !WITH_EARLY_DATA(conn))
+            || (message_type == END_OF_EARLY_DATA)) {
+        K_send(conn, S2N_HANDSHAKE_SECRET);
```

Comment:
We store all 3 base / extract secrets, but not all 5 traffic secrets. We could store all of them (and do store 2 atm), but I don't think we'd see any real benefit. For example, since this is the client, waiting for the end of early data doesn't add a round trip.

If we decide to store more secrets, it would be a fairly lightweight change. The derive method would just need to look like the extract method, and return the stored traffic secret if available.